### PR TITLE
patch: update currency based on location

### PR DESCRIPTION
**Pull request by tinashechiraya**
Description

The user will be prompted to allow access to location so product price and currency can be adjusted according to users location

Adjustments:

- fetch user location
- send currency to back-end
- using currency from fronted fetch the appropriate values from db

### CODE REVIEW

**Comment by dimasciput** [comment]
@tinashechiraya 

Let's avoid fetching another API here. Instead, we can change the API that retrieves product details to POST request. The React frontend can send the user's location (if available) to the server, and the server will handle pricing logic and return the products with the correct prices. 

For example, the current API to get product details:

https://geohosting.dev.do.kartoza.com/api/products/G3W/

Let's convert this to a POST request, with the frontend adding the user's location if it exists.

File: django_project/geohosting/serializer/product.py [hunk 1]
```diff
@@ -34,9 +34,19 @@ class ProductDetailSerializer(serializers.ModelSerializer):
     packages = serializers.SerializerMethodField()
     product_meta = serializers.SerializerMethodField()
 
+    def __init__(self, *args, **kwargs):
+        self.currency = kwargs.pop('currency', None)
+        super().__init__(*args, **kwargs)
+
     def get_packages(self, obj: Product):
         packages = obj.packages.all()
```
Comment (meomancer) [inline django_project/geohosting/serializer/product.py:42]: Hi @tinashechiraya 
You can just filter it here, so we make sure it is the correct currency
By default is ZAR if no currency shown

File: django_project/geohosting/serializer/product.py [hunk 2]
```diff
@@ -34,9 +34,19 @@ class ProductDetailSerializer(serializers.ModelSerializer):
...
-        preferred_currency_order = ['EUR', 'USD', 'ZAR']
+
+        preferred_currency_order = (
```
Comment (meomancer) [inline django_project/geohosting/serializer/product.py:44]: We can delete preferred currency order as we already know what currency should be returned
And also we can delete 
```
unique_packages = {}
```
code, as we just return package with correct currency

**Comment by meomancer** [comment]
Thank you @tinashechiraya 
Please create test for this

File: django_project/geohosting/src/redux/reducers/productsSlice.ts [hunk 3]
```diff
@@ -52,6 +52,23 @@ const initialState: ProductsState = {
   detailError: null,
 };
 
+const getCurrencyBasedOnLocation = async () => {
```
Comment (meomancer) [inline django_project/geohosting/src/redux/reducers/productsSlice.ts:55]: Hi @tinashechiraya 
Could put this function to new utils file?
In this folder: https://github.com/kartoza/GeoHosting/tree/main/django_project/geohosting/src/utils
So we can pick it up faster

**localtion_helper.tsx** for example

File: django_project/geohosting/tests/test_product.py [hunk 4]
```diff
@@ -207,3 +209,72 @@ def test_fetch_products_image_download_fail(self, mock_get, mock_fetch_erpnext_d
...
+        response = self.client.get(self.url, {'currency': 'USD'})
+        self.assertEqual(response.status_code, 200)
+
+        # Check if serializer is returning correct packages based on currency
+        product_data = response.data
+        serializer = ProductDetailSerializer(instance=self.product, context={'currency': 'USD'})
+
+        self.assertEqual(product_data['packages'], serializer.data['packages'])
+        self.assertEqual(product_data['images'], serializer.data['images'])
+        self.assertEqual(product_data['product_meta'], serializer.data['product_meta'])
```
Comment (meomancer) [inline django_project/geohosting/tests/test_product.py:267]: Please also assert equal the currency of the package
self.assertEqual(product_data['packages'][0]['currency'], 'USD')

**Comment by meomancer** [comment]
@tinashechiraya 
Please split the function:
- Returning location
- Returning currency

**Comment by meomancer** [comment]
LGTM Thank you @tinashechiraya

**Pull request merged**